### PR TITLE
Check bitcoind version at startup

### LINKED ISSUE #324: Check bitcoind version at startup
`bitcoind`'s releases are versioned. New releases bring new features. `revaultd` depends on the newest features of `bitcoind`. Checking the version at startup is for sanity checking that the features we depend on are available on the `bitcoind` we are configured to talk to.

### PR DESCRIPTION
**Pull request by meryacine**
Fixes #324
Closes #378

### CODE REVIEW

**Comment by darosior** [comment]
Please make sure your PR passes CI.  Also, please do not duplicate PRs. This is a duplicate of #385.

With that being said, the code here seems closer to the goal than that of #385. Maybe you could co-operate in a single PR?

**Comment by darosior** [comment]
@meryacine given the lack of response of Gourab in #378, feel free to move forward with this PR.

**Comment by meryacine** [comment]
@darosior I think that was some kind of a bug or I did something wrong while force pushing to resolve the conflicts with master with a rebase, the PR auto closed itself :/. Reopening it.

**Comment by meryacine** [comment]
@darosior I think this test failed by coincidence (maybe a tight timeout?).

I ran `tests/test_spend.py::test_large_spends` manually on a cloud VM and passed (took `1791.47s (0:29:51)`).

**Comment by darosior** [comment]
`test_large_spend` is flaky at the moment so no worries if it spuriously fails. I'll have a look :)
------- Original Message -------

File: src/bitcoind/mod.rs [hunk 1]
```diff
@@ -28,6 +28,9 @@ use jsonrpc::{
 /// A retry happens every 1 second, this makes us give up after one minute.
 const BITCOIND_RETRY_LIMIT: usize = 60;
 
+/// The minimum bitcoind version that can be used with revaultd.
+const MIN_BITCOIND_VERSION: u64 = 210000;
```
Comment (darosior) [inline src/bitcoind/mod.rs:32]: We are using some features from 22.0, so can you make it `220000`?

File: src/bitcoind/mod.rs [hunk 2]
```diff
@@ -113,13 +116,34 @@ fn check_bitcoind_network(
...
 }
 
+fn check_bitcoind_version(bitcoind: &BitcoinD) -> Result<(), BitcoindError> {
+    let network_info = bitcoind.getnetworkinfo()?;
+    let bitcoind_version = network_info
+        .get("version")
+        .and_then(|v| v.as_u64())
+        .ok_or_else(|| {
+            BitcoindError::Custom("No valid 'version' in getnetworkinfo response?".to_owned())
+        })?;
```
Comment (darosior) [inline src/bitcoind/mod.rs:126]: I think we should just panic in this case.
```suggestion
        .expect("No valid 'version' in getnetworkinfo response?");
```

**Review by darosior — COMMENTED** [review]
Looks good to me. I just have 2 small comments.

**Comment by darosior** [comment]
Note: i've edited the OP to add a link that will automatically close the issue fixed by this PR.

File: src/bitcoind/mod.rs (hunk 1 above)
Comment (meryacine) [reply src/bitcoind/mod.rs:32]: Ah sure, we will need to also change [this](https://github.com/revault/revaultd/blob/master/README.md#:~:text=(version%200.21%20minimum)) then.

Comment (darosior) [reply src/bitcoind/mod.rs:32]: Oh, yes, good catch!

File: src/bitcoind/mod.rs (hunk 2 above)
Comment (meryacine) [reply src/bitcoind/mod.rs:126]: I saw a similar error handling in [here](https://github.com/revault/revaultd/blob/fd54a81111c2450530de4879de6318110a3869e7/src/bitcoind/mod.rs#L97) and assumed that bitcoind might not respond also for the case of propping for version if it is warming up. If this is not the case then we can safely panic.

Comment (darosior) [reply src/bitcoind/mod.rs:126]: Heh, no. You are right. Let's keep it like that for now.

**Review by darosior — APPROVED** [review]
Pleasant to read a neat and well organized PR! Thanks for fixing this issue.

utACK 0294c5ab19a277d18d4ad0041dfbfdae38312c5f

**Pull request merged**